src/linear_algebra/clifford_algebra.lean

```diff
@@ -0,0 +1,156 @@
+
+The canonical linear map `M → clifford_algebra Q` is denoted `clifford_algebra.ι Q`.
+
+## Theorems
+
+The main theorems proved ensure that `clifford_algebra Q` satisfies the universal property
+of the Clifford algebra.
+1. `ι_comp_lift` is the fact that the composition of `ι Q` with `lift Q f cond` agrees with `f`.
+2. `lift_unique` ensures the uniqueness of `lift Q f cond` with respect to 1.
+
+Additionally, when `Q = 0` an `alg_equiv` to the `exterior_algebra` is provided as `as_exterior`.
+
+## Implementation details
+
+The Clifford algebra of `M` is constructed as a quotient of the tensor algebra, as follows.
+1. We define a relation `clifford_algebra.rel Q` on `tensor_algebra R M`.
+   This is the smallest relation which identifies squares of elements of `M` with `Q m`.
+2. The Clifford algebra is the quotient of the tensor algebra by this relation.
+
+This file is almost identical to `linear_algebra/exterior_algebra.lean`.
+-/
+
+variables {R : Type*} [comm_ring R]
+variables {M : Type*} [add_comm_group M] [module R M]
+variables (Q : quadratic_form R M)
+
+variable {n : ℕ}
+
+namespace clifford_algebra
+open tensor_algebra
+
+/-- `rel` relates each `ι m * ι m`, for `m : M`, with `Q m`.
+
+The Clifford algebra of `M` is defined as the quotient modulo this relation.
+-/
+inductive rel : tensor_algebra R M → tensor_algebra R M → Prop
+| of (m : M) : rel (ι R m * ι R m) (algebra_map R _ (Q m))
+
+end clifford_algebra
+
+/--
+The Clifford algebra of an `R`-module `M` equipped with a quadratic_form `Q`.
+-/
+@[derive [inhabited, semiring, algebra R]]
+def clifford_algebra := ring_quot (clifford_algebra.rel Q)
+
+namespace clifford_algebra
+
+/--
+The canonical linear map `M →ₗ[R] clifford_algebra Q`.
+-/
+def ι : M →ₗ[R] clifford_algebra Q :=
+(ring_quot.mk_alg_hom R _).to_linear_map.comp (tensor_algebra.ι R)
+
+/-- As well as being linear, `ι m` squares to the quadratic form -/
+@[simp]
+theorem ι_square_scalar (m : M) : ι Q m * ι Q m = algebra_map R _ (Q m) :=
+begin
+  erw [←alg_hom.map_mul, ring_quot.mk_alg_hom_rel R (rel.of m), alg_hom.commutes],
+  refl,
+end
+
+variables {A : Type*} [semiring A] [algebra R A]
+
+/--
+Given a linear map `f : M →ₗ[R] A` into an `R`-algebra `A`, which satisfies the condition:
+`cond : ∀ m : M, f m * f m = Q(m)`, this is the canonical lift of `f` to a morphism of `R`-algebras
+from `clifford_algebra Q` to `A`.
+-/
+def lift (f : M →ₗ[R] A) (cond : ∀ m, f m * f m = algebra_map _ _ (Q m)) :
+  clifford_algebra Q →ₐ[R] A :=
+ring_quot.lift_alg_hom R (tensor_algebra.lift R f)
+  (λ x y h, by {
+    induction h,
+    rw [alg_hom.commutes, alg_hom.map_mul, tensor_algebra.lift_ι_apply, cond], })
+
+variables {Q}
+
+@[simp]
+theorem ι_comp_lift (f : M →ₗ[R] A) (cond : ∀ m, f m * f m = algebra_map _ _ (Q m)) :
+  (lift Q f cond).to_linear_map.comp (ι Q) = f :=
+by { ext, simp [lift, ι] }
+
+@[simp]
+theorem lift_ι_apply (f : M →ₗ[R] A) (cond : ∀ m, f m * f m = algebra_map _ _ (Q m)) (x) :
```

Comment:
Since we define `ι_comp_lift` the same as `lift_ι_apply` but with an extra `ext` in front, wouldn't it be nicer if `lift_ι_apply` was defined first and `ι_comp_lift := by { ext, apply lift_ι_apply }`?

Reply:
Good point - that also raises that one can be moved after the reducibility barrier, since it follows trivially from the other.

Again, this sounds like something for the follow-up PR touching all four files.